### PR TITLE
Rewrite the frame parser to remove yield from and put back Python 2.7 support

### PR DESCRIPTION
**Pull request by jeamland**
Addresses #22

### CODE REVIEW

**Comment by njsmith** [comment]
> @jeamland requested your review on this pull request. 

Definitely want to look at this, but I'm at PyCon and talking at the language summit tomorrow so you might need to ping me in a few days if I forget :-)

**Comment by jeamland** [comment]
I'll just come down there and annoy you myself. =)

**Comment by njsmith** [comment]
Wait, are you here?

**Comment by jeamland** [comment]
Not yet. I'm heading down from Seattle on Thursday.

File: wsproto/utf8validator.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+                As soon as an octet is encountered which renders the octet sequence
+                invalid, a quad with ``valid? == False`` is returned. ``currentIndex`` returns
+                the index within the currently consumed chunk, and ``totalIndex`` the
+                index within the total consumed sequence that was the point of bail out.
+                When ``valid? == True``, currentIndex will be ``len(ba)`` and ``totalIndex`` the
+                total amount of consumed bytes.
+                """
+                #
+                # The code here is written for optimal JITting in PyPy, not for best
+                # readability by your grandma or particular elegance. Do NOT touch!
```
Comment (njsmith) [inline wsproto/utf8validator.py]: Okay, actually yeah I think we should touch this comment. My mom will become a grandma in a few months, and my first job in high school was adding new features to old COBOL systems she'd written before I was born. They were still running businesses. In the mean time she went back to school to get a masters in education, and had a second career teaching art/STEM/coding. She learned Python last year, after retiring. This shot at grandmas is misogynistic garbage, and there's absolutely no reason for us to propagate it further where more people might encounter it.

Suggested text: "The code here is written for optimal JITtin in PyPy, not for best readability or particular elegance. Do NOT touch!"

File: wsproto/utf8validator.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
```
Comment (njsmith) [inline wsproto/utf8validator.py]: See above

File: wsproto/utf8validator.py (hunk 1 above)
Comment (jeamland) [inline wsproto/utf8validator.py]: Gawd, 100% agree. If I'd actually noticed this I'd've changed it straight up.

File: wsproto/utf8validator.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+###############################################################################
+
+# Note: This code is a Python implementation of the algorithm
+# "Flexible and Economical UTF-8 Decoder" by Bjoern Hoehrmann
+# bjoern@hoehrmann.de, http://bjoern.hoehrmann.de/utf-8/decoder/dfa/
```
Comment (jeamland) [inline wsproto/utf8validator.py]: Is there a test suite? I can't find it in the autobahn code where I nicked this from... I'll have more of a look.

Comment (njsmith) [inline wsproto/utf8validator.py]: Oh ugh, they really don't have one do they. All I can find is in the `wsaccel` package there's 1 lousy test (just `"a bunch of unicode".encode("utf-8")` and then asserting it's valid), and `autobahn-python` has none :-(

Comment (njsmith) [inline wsproto/utf8validator.py]: Ok, looking again, maybe it's not so bad – the [uncovered code that was bothering me](https://codecov.io/gh/python-hyper/wsproto/src/rewrite-the-third/wsproto/utf8validator.py) appears to be (a) the python 3 version that we never use, and (b) the `decode` method that we never use. The py2 `validate` method that we care about is actually covered. I'm not really sure how meaningful that is given that the relevant state space is this opaque DFA thing that coverage can't see, but we know the autobahn protocol tests do exercise some reasonable amount of weird utf8 edge cases (certainly enough to have alerted us to the problem in the first place!).

It does seem like we might want to chop out the code we aren't using to fix the `six` issue and make codecov happier?

File: wsproto/frame_protocol.py [hunk 4]
```diff
@@ -9,36 +9,43 @@
...
 # RFC6455, Section 5.2 - Base Framing Protocol
 
 # Payload length constants
 PAYLOAD_LENGTH_TWO_BYTE = 126
 PAYLOAD_LENGTH_EIGHT_BYTE = 127
 MAX_PAYLOAD_NORMAL = 125
-MAX_PAYLOAD_TWO_BYTE = 2 ** 16
-MAX_PAYLOAD_EIGHT_BYTE = 2 ** 64
+MAX_PAYLOAD_TWO_BYTE = 2 ** 16 - 1
+MAX_PAYLOAD_EIGHT_BYTE = 2 ** 64 - 1
```
Comment (jeamland) [inline wsproto/frame_protocol.py]: We're lacking in send-side tests at the moment.

File: wsproto/frame_protocol.py [hunk 5]
```diff
@@ -139,68 +146,169 @@ def _truncate_utf8(data, nbytes):
...
+
+    def consume_exactly(self, nbytes):
+        if len(self.buffer) - self.bytes_used < nbytes:
+            return None
+
+        return self.consume_at_most(nbytes)
+
+    def commit(self):
+        # In CPython 3.4+, del[:n] is amortized O(n), *not* quadratic
+        del self.buffer[:self.bytes_used]
```
Comment (jeamland) [inline wsproto/frame_protocol.py]: Would `self.buffer = self.buffer[self.bytes_used:]` be better?

Comment (njsmith) [inline wsproto/frame_protocol.py]: No, that would make it quadratic everywhere :-)

If we are worried then I suggest we leave it like it is in this PR and file an issue so we don't forget about it. h11 has some code we can steal that implements the amortized `O(n)` behavior but it's a little complicated.

**Comment by jeamland** [comment]
@njsmith I think I've addressed everything. If there's anything else let me know otherwise I'd like to merge this.

**Comment by njsmith** [comment]
Awesome, thanks

**Comment by agronholm** [comment]
Much appreciated!
